internal/controller/warehouses/git.go

```diff
@@ -231,6 +272,72 @@ func ignores(tagName string, ignore []string) bool {
 	return false
 }
 
+// pathsFilterPositive returns true when ScanPaths and/or IgnorePaths
+// filters match one or more commit diffs and new Freight is
+// to be produced. It returns false otherwise.
+func matchesPathsFilters(scanPaths []string, ignorePaths []string, diffs []string) (bool, error) {
+	filteredDiffs := make([]string, 0, len(diffs))
+	for _, diffPath := range diffs {
+		// matchesScanPaths case is a bit different from matchesIgnorePaths
+		// in the way that if scanPaths string array is empty - it matches
+		// ANY change so we need to have a check for that
+		var matchesScanPaths bool
+		if scanPaths == nil {
+			matchesScanPaths = true
+		} else {
+			var err error
+			matchesScanPaths, err = matchesRegexList(diffPath, scanPaths)
+			if err != nil {
+				return false, errors.Wrapf(
+					err,
+					"error checking scanPaths regexps match against path %q",
+					diffPath,
+				)
+			}
+		}
+		matchesIgnorePaths, err := matchesRegexList(diffPath, ignorePaths)
+		if err != nil {
+			return false, errors.Wrapf(
+				err,
+				"error checking ignorePaths regexps match against path %q",
+				diffPath,
+			)
+		}
+		// combined filter decision, positive for matching scanPaths and
+		// unmatching ignorePaths
+		if matchesScanPaths && !matchesIgnorePaths {
+			filteredDiffs = append(filteredDiffs, diffPath)
+		}
+	}
+	if len(filteredDiffs) > 0 {
+		return true, nil
+	}
+	return false, nil
+}
+
+// matchesRegexList is a general purpose function iterating given slice of regexp
+// strings (regexList) to check if any of them match the stringToMatch string, if
+// a regexp fails to compile it returns false and error, if match is found it
+// returns true and if match is not found it returns false.
+func matchesRegexList(stringToMatch string, regexList []string) (bool, error) {
```

Comment:
This is quite a costly function, as recompilation of regular expressions happens repeatedly for every `diffPath`, while `scanPaths` and `ignorePaths` are static.

Given this, I would restructure the code in such a way that the lists of regular expressions are compiled once before attempting to match them against diff paths.

Reply:
agree